testtools/UART_interface/esp_arduino_uart_interface.py

```diff
@@ -0,0 +1,164 @@
+        print("Failed to connect to saved WiFi network.")
+        azure_test_firmware_errors.SDK_ERRORS += 1
+
+
+# ------- interface class -------
+class esp_uart_interface(uart_interface):
+
+    message_callbacks = 0
+    messages_sent = 5
+
+    def check_sample_errors(self, line):
+        if "Confirm Callback" in line:
+            self.message_callbacks += 1
+
+    # If there is a sudden disconnect, program should report line in input script reached, and close files.
+    # method to write to serial line with connection monitoring
+    def serial_write(self, ser, message, file=None):
+
+        # Check that the device is no longer sending bytes
+        if ser.in_waiting:
+            self.serial_read(ser, message, file)
+
+        # Check that the serial connection is open
+        if ser.writable():
+            wait = serial_settings.mxchip_buf_pause  # wait for at least 50ms between 128 byte writes.
+            buf = bytearray((message.strip() + '\r\n').encode('ascii'))
+            buf_len = len(buf)  # needed for loop as buf is a destructed list
+            bytes_written = 0
+            timeout = time.time()
+
+            while bytes_written < buf_len:
+                temp_written = ser.write(buf[:128])
+                buf = buf[temp_written:]
+                bytes_written += temp_written
+                time.sleep(wait)
+                if (time.time() - timeout > serial_settings.serial_comm_timeout):
+                    break
+
+            return bytes_written
+        else:
+            try:
+                time.sleep(2)
+                ser.open()
+                self.serial_write(ser, message, file)
+            except:
+                return False
+
+    # Read from serial line with connection monitoring
+    # If there is a sudden disconnect, program should report line in input script reached, and close files.
+    def serial_read(self, ser, message, file, first_read=False):
+
+        # Special per opt handling:
+        if "exit" in message and first_read:
+            time.sleep(serial_settings.wait_for_flash)
+            output = ser.in_waiting
+            while output < 4:
+                time.sleep(1)
+                output = ser.in_waiting
+                print("%d bytes in waiting" % output)
+
+        if ser.readable():
+            output = ser.readline(ser.in_waiting)
+            output = output.decode(encoding='utf-8', errors='ignore')
+
+            check_firmware_errors(output)
+            check_sdk_errors(output)
+            self.check_sample_errors(output)
+            print(output)
+            try:
+                # File must exist to write to it
+                file.writelines(output)
+            except:
+                pass
+
+            return output
+        else:
+            try:
+                time.sleep(2)
+                ser.open()
+                self.serial_read(ser, message, file, first_read=True)
+            except:
+                return False
```

Comment:
it seems odd to have a function return a boolean and a string. Consider return `None` instead of false in the fail case. Here and elsewhere.